File: source/rust/src/worker.rs
// Worker module: hot loop performance-critical code

use std::cell::RefCell;

#[allow(unused_imports)]
use crate::loops::Loops;
#[allow(unused_imports)]
use crate::tarjan::Tarjan;

// Select Tarjan or Loops
pub(crate) type Relations = Tarjan;

fn swap(arr: &mut Vec<usize>, i: usize, j: usize) {
    let temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

// Heap's algorithm: tally all perms with a fixed length k prefix
fn tally_perms<F>(perm: &mut Vec<usize>, k: usize, mut work: F)
where
    F: FnMut(&mut Vec<usize>),
{
    let n = perm.len();

    fn generate<F>(perm: &mut Vec<usize>, j: usize, k: usize, work: &mut F)
    where
        F: FnMut(&mut Vec<usize>),
    {
        if j < k {
            work(perm);
        } else {
            generate(perm, j - 1, k, work);

            for i in k..j {
                if (j - k) % 2 == 0 {
                    swap(perm, k, j);
                } else {
                    swap(perm, i, j);
                }
                generate(perm, j - 1, k, work);
            }
        }
    }

    generate(perm, n - 1, k, &mut work);
}

// Count cycles in a signed permutation
fn count_cycles(n: usize, perm: &Vec<usize>, signs: usize, rel: &mut Relations) -> usize {
    rel.reset(2 * n);

    for i in 0..n {
        let j = perm[i];

        if ((signs >> i) & 1) == 1 {
            rel.unite(i, j + n);
            rel.unite(i + n, j);
        } else {
            rel.unite(i, j);
            rel.unite(i + n, j + n);
        }
    }

    rel.set_count()
}

// Tally cycle counts across all signs for one perm
fn tally_signs(n: usize, perm: &mut Vec<usize>, tally: &mut Vec<i64>, rel: &mut Relations) {
    let max_bits = 1 << n;

    for signs in 0..max_bits {
        let cycles = count_cycles(n, perm, signs, rel);
        let index = 2 * n - cycles;
        tally[index] += 1;
    }
}

// Thread-local storage for Relations and tally to avoid repeated allocations
thread_local! {
    static THREAD_DATA: RefCell<(Relations, Vec<i64>)> =
        RefCell::new((Relations::create(48), vec![0; 48]));
}

// Process one parcel: tally all cycle counts with given prefix
pub(crate) fn run_prefix(n: usize, k: usize, prefix: Vec<usize>) -> Vec<i64> {
    THREAD_DATA.with(|data| {
        let mut data = data.borrow_mut();
        let (rel, tally) = &mut *data;

        if tally.len() < 2 * n {
            tally.resize(2 * n, 0);
        }
        for i in 0..2 * n {
            tally[i] = 0;
        }

        let mut perm = prefix;
        let work = |perm: &mut Vec<usize>| {
            tally_signs(n, perm, tally, rel);
        };

        tally_perms(&mut perm, k, work);

        tally[0..2 * n].to_vec()
    })
}

File: source/ocaml/src/worker.ml
(* Worker module: hot loop performance-critical code *)

(* Select Tarjan or Loops *)
module Sets = Tarjan let name = "Tarjan"

(* Heap's algorithm: tally all perms with a fixed length k prefix *)
let tally_perms perm k work =
  let n = Array.length perm in
  let swap = Parallel.swap in
  let rec generate j =
    if j < k then work perm
    else
      ( generate (j - 1);
        for i = k to j - 1 do
          if (j - k) mod 2 = 0 then swap perm k j else swap perm i j;
          generate (j - 1)
        done )
  in
  generate (n - 1)

(* Count cycles in a signed permutation *)
let count_cycles n perm signs rel =
  Sets.reset rel (2 * n);
  for i = 0 to n - 1 do
    let j = perm.(i) in
    if (signs lsr i) land 1 = 1 then
      ( Sets.unite rel i (j + n);
        Sets.unite rel (i + n) j )
    else
      ( Sets.unite rel i j;
        Sets.unite rel (i + n) (j + n) )
  done;
  Sets.set_count rel

(* Tally cycle counts across all signs for one perm *)
let tally_signs n perm tally rel =
  let maxBits = 1 lsl n in
  for signs = 0 to maxBits - 1 do
    let cycles = count_cycles n perm signs rel in
    let index = (2 * n) - cycles in
    tally.(index) <- tally.(index) + 1
  done

(* Process one parcel: tally all cycle counts with given prefix *)
let run_prefix n k perm =
  let perm = Array.of_list perm in
  let tally = Array.make (2 * n) 0 in
  let rel = Sets.create (2 * n) in
  let work perm = tally_signs n perm tally rel in
  tally_perms perm k work;
  tally
